tests/conftest.py

```diff
@@ -360,3 +364,51 @@ def message_parameters_dict():
             "model_text": MESSAGE_BLOB,
         },
     }
+
+
+@contextmanager
+def temp_environ(**vars):
+    """
+    A simple context manager for temporarily setting environment variables
+
+    Kwargs:
+        variables to be set and their values
+
+    Returns:
+        None
+    """
+    from os import environ
+    original = dict(environ)
+    os.environ.update(vars)
+    try:
+        yield  # no value needed
+    finally:
+        # restore original data
+        os.environ.clear()
+        os.environ.update(original)
+
+
+@contextmanager
+def temp_config(data, perms=stat.S_IRUSR | stat.S_IWUSR):
+    """
+    A context manager which creates a temporary config file with specified data and permissions
+
+    Args:
+        data: the data to be written to the file
+        perms: the permissions which should be set on the file.
+            The default value is to use the standard, safe permissions
+
+    Returns:
+        None
+    """
+    config_path = configure.get_config_path()
```

Comment:
It feels dangerous that this function will delete my config if I invoke it wrong (that is, without overriding `XDG_CONFIG_HOME`).

Maybe this should always write into a temporary directory, and then yield that temporary directory path?

Reply:
This function has no latitude to choose the location, as the code under test will always just invoke `configure.get_config_path()`. However, your point about this being dangerous if someone writes a bad test is a good one. I think we can defend against that by having this function check that `XDG_CONFIG_HOME` is at least set, as any test which is not doing so is almost certainly wrong. That could still do bad things if there is both a bad test and the user keeps `XDG_CONFIG_HOME` set in their environment; I will see if we can reasonably distinguish pytest's temporary directory paths from others, so it can check that the variable is both set and points to a temporary directory.

Reply:
> This function has no latitude to choose the location, as the code under test will always just invoke `configure.get_config_path()`

I was picturing flipping the flow you have in the current tests. Right now, you
 - set XDG_CONFIG_HOME to a temporary directory
 - call this function to write a config file into XDG_CONFIG_HOME
 - yield execution to the actual test, which is none the wiser that it's operating against a test config in a temp directory.
 
 But you could
 - call this function to write a config file into a temporary directory, and yield it
 - set XDG_CONFIG_HOME to whatever this function yielded
 - yield execution to the actual test, which is none the wiser that it's operating against a test config in a temp directory.
 
 Would that work?

Reply:
That should be possible; I'll try it out. I think `tmpdir` may have to be passed into `temp_config` explicitly then, unfortunately, as pytest seems to be very tight-fisted with information about where it is (I think because it's trying to make it per-test, to support concurrent tests).